{
  if (cancelled) {
    return;
  }
  cancelled=true;
  Exception failure=null;
  Runnable runnable=callbacks.poll();
  while (runnable != null) {
    try {
      runnable.run();
    }
 catch (    Exception ex) {
      if (failure != null) {
        Throwable lastEx=ex;
        while (lastEx.getCause() != null) {
          lastEx=lastEx.getCause();
        }
        lastEx.initCause(failure);
      }
      failure=ex;
    }
    runnable=callbacks.poll();
  }
  if (failure != null) {
    throw new RuntimeException(failure);
  }
}
